[Bug]: Infinite recursion when registering/resolving data types that depend on each other

### What happened?

In https://github.com/hdmf-dev/hdmf-common-schema/pull/91:
the DynamicTable data type contains a collection of MeaningsTable objects.
MeaningsTable is a subtype of a DynamicTable.

When loading the namespace, we get a recursion error below. This is not an issue if the collection of MeaningsTable objects is added to a child type of DynamicTable, e.g., a new type DynamicTableWithMeanings, but this is not ideal.

I think this schema, where A contains B and B is a subtype of A (B includes A), should be allowed, though it may be tricky to implement. I think we should also allow the simpler case where A contains A (https://github.com/hdmf-dev/hdmf/issues/794). I think both HDMF and MatNWB will need to be updated to handle these use cases. @oruebel @bendichter @ehennestad thoughts?

The refactor of spec resolution in https://github.com/hdmf-dev/hdmf/pull/1326/changes in src/hdmf/spec/namespace.py lines 445-449 will also need to handle this.

### Steps to Reproduce

```python
Use that branch of hdmf-common and run tests
```

### Traceback

```pytb
tests/unit/spec_tests/test_spec_catalog.py:4: in <module>
    from hdmf.testing import TestCase
src/hdmf/testing/__init__.py:1: in <module>
    from .testcase import TestCase, H5RoundTripMixin
src/hdmf/testing/testcase.py:10: in <module>
    from ..common import validate as common_validate, get_manager
src/hdmf/common/__init__.py:268: in <module>
    load_namespaces(__resources['namespace_path'])
src/hdmf/utils.py:596: in func_call
    return func(**pargs)
           ^^^^^^^^^^^^^
src/hdmf/common/__init__.py:136: in load_namespaces
    return __TYPE_MAP.load_namespaces(namespace_path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
src/hdmf/utils.py:592: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
src/hdmf/build/manager.py:491: in load_namespaces
    deps = self.__ns_catalog.load_namespaces(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
src/hdmf/utils.py:592: in func_call
    return func(args[0], **pargs)
           ^^^^^^^^^^^^^^^^^^^^^^
src/hdmf/spec/namespace.py:585: in load_namespaces
    ret[ns['name']] = self.__load_namespace(ns, r, resolve=resolve)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
src/hdmf/spec/namespace.py:473: in __load_namespace
    self.__register_type(ndt, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:496: in __register_type
    self.__register_dependent_types(spec, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:525: in __register_dependent_types
    __register_dependent_types_helper(child_spec, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:519: in __register_dependent_types_helper
    __register_dependent_types_helper(child_spec, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:512: in __register_dependent_types_helper
    self.__register_type(spec.data_type_inc, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:496: in __register_type
    self.__register_dependent_types(spec, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:522: in __register_dependent_types
    self.__register_type(spec.data_type_inc, inc_ns, catalog, registered_types)
src/hdmf/spec/namespace.py:496: in __register_type
    self.__register_dependent_types(spec, inc_ns, catalog, registered_types)
E   RecursionError: maximum recursion depth exceeded
!!! Recursion detected (same locals & position)
```

### Operating System

macOS

### Python Version

3.13

### Package Versions

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 
 ### Changed
 - Added a collapsible "columns" section to the `DynamicTable` HTML representation that displays column descriptions, making it easier to inspect column metadata in notebooks. @h-mayorquin [#1369](https://github.com/hdmf-dev/hdmf/pull/1369)
+- Added ability to create specs with circular dependencies, e.g., Type A contains Type B, and Type B extends Type A, or Type A contains a reference to Type A (self-reference). @rly [#1374](https://github.com/hdmf-dev/hdmf/pull/1374)
 
 ### Fixed
 

--- a/src/hdmf/spec/namespace.py
+++ b/src/hdmf/spec/namespace.py
@@ -490,12 +490,11 @@ class NamespaceCatalog:
     def __register_type(self, ndt, inc_ns, catalog, registered_types, in_progress_registrations):
         """Register a type and its dependencies from a namespace into a catalog.
 
-        Args:
-            ndt: The name of the data type to register
-            inc_ns: The namespace containing the type
-            catalog: The catalog to register the type into
-            registered_types: Set of already registered types (to avoid re-registering)
-            in_progress_registrations: Set of types currently being registered (for circular dependency detection)
+        :param ndt: The name of the data type to register
+        :param inc_ns: The namespace containing the type
+        :param catalog: The catalog to register the type into
+        :param registered_types: Set of already registered types (to avoid re-registering)
+        :param in_progress_registrations: Set of types currently being registered (for circular dependency detection)
         """
         if ndt in registered_types or ndt in in_progress_registrations:
             # Already registered or currently being registered (circular dependency)

--- a/src/hdmf/spec/namespace.py
+++ b/src/hdmf/spec/namespace.py
@@ -470,8 +470,7 @@ class NamespaceCatalog:
                     types_to_load = inc_ns.get_registered_types()  # load all types in namespace
                 registered_types = set()
                 for ndt in types_to_load:
-                    in_progress_registrations = set()
-                    self.__register_type(ndt, inc_ns, catalog, registered_types, in_progress_registrations)
+                    self.__register_type(ndt, inc_ns, catalog, registered_types)
                 included_types[s['namespace']] = tuple(sorted(registered_types))
             else:
                 raise ValueError("Spec '%s' schema must have either 'source' or 'namespace' key" % ns_name)
@@ -487,15 +486,18 @@ class NamespaceCatalog:
                                             catalog=catalog)
         return included_types
 
-    def __register_type(self, ndt, inc_ns, catalog, registered_types, in_progress_registrations):
+    def __register_type(self, ndt, inc_ns, catalog, registered_types, in_progress_registrations=None):
         """Register a type and its dependencies from a namespace into a catalog.
 
         :param ndt: The name of the data type to register
         :param inc_ns: The namespace containing the type
         :param catalog: The catalog to register the type into
         :param registered_types: Set of already registered types (to avoid re-registering)
-        :param in_progress_registrations: Set of types currently being registered (for circular dependency detection)
+        :param in_progress_registrations: Set of types currently being registered (for circular dependency detection).
+            None if the registration process is starting.
         """
+        if in_progress_registrations is None:  
+            in_progress_registrations = set()
         if ndt in registered_types or ndt in in_progress_registrations:
             # Already registered or currently being registered (circular dependency)
             return

--- a/src/hdmf/spec/namespace.py
+++ b/src/hdmf/spec/namespace.py
@@ -470,7 +470,8 @@ class NamespaceCatalog:
                     types_to_load = inc_ns.get_registered_types()  # load all types in namespace
                 registered_types = set()
                 for ndt in types_to_load:
-                    self.__register_type(ndt, inc_ns, catalog, registered_types)
+                    in_progress_registrations = set()
+                    self.__register_type(ndt, inc_ns, catalog, registered_types, in_progress_registrations)
                 included_types[s['namespace']] = tuple(sorted(registered_types))
             else:
                 raise ValueError("Spec '%s' schema must have either 'source' or 'namespace' key" % ns_name)
@@ -486,43 +487,71 @@ class NamespaceCatalog:
                                             catalog=catalog)
         return included_types
 
-    def __register_type(self, ndt, inc_ns, catalog, registered_types):
-        if ndt in registered_types:
-            # already registered
-            pass
-        else:
-            spec = inc_ns.get_spec(ndt)
-            spec_file = inc_ns.catalog.get_spec_source_file(ndt)
-            self.__register_dependent_types(spec, inc_ns, catalog, registered_types)
-            if isinstance(spec, DatasetSpec):
-                built_spec = self.dataset_spec_cls.build_spec(spec)
-            else:
-                built_spec = self.group_spec_cls.build_spec(spec)
-            registered_types.add(ndt)
-            catalog.register_spec(built_spec, spec_file)
+    def __register_type(self, ndt, inc_ns, catalog, registered_types, in_progress_registrations):
+        """Register a type and its dependencies from a namespace into a catalog.
 
-    def __register_dependent_types(self, spec, inc_ns, catalog, registered_types):
-        """Ensure that classes for all types used by this type are registered
+        Args:
+            ndt: The name of the data type to register
+            inc_ns: The namespace containing the type
+            catalog: The catalog to register the type into
+            registered_types: Set of already registered types (to avoid re-registering)
+            in_progress_registrations: Set of types currently being registered (for circular dependency detection)
         """
-        # TODO test cross-namespace registration...
-        def __register_dependent_types_helper(spec, inc_ns, catalog, registered_types):
-            if isinstance(spec, (GroupSpec, DatasetSpec)):
-                if spec.data_type_inc is not None:
-                    # TODO handle recursive definitions
-                    self.__register_type(spec.data_type_inc, inc_ns, catalog, registered_types)
-                if spec.data_type_def is not None:  # nested type definition
-                    self.__register_type(spec.data_type_def, inc_ns, catalog, registered_types)
-            else:  # spec is a LinkSpec
-                self.__register_type(spec.target_type, inc_ns, catalog, registered_types)
-            if isinstance(spec, GroupSpec):
-                for child_spec in (spec.groups + spec.datasets + spec.links):
-                    __register_dependent_types_helper(child_spec, inc_ns, catalog, registered_types)
+        if ndt in registered_types or ndt in in_progress_registrations:
+            # Already registered or currently being registered (circular dependency)
+            return
+        # Track that we're currently registering this type
+        in_progress_registrations.add(ndt)
+        spec = inc_ns.get_spec(ndt)
+        spec_file = inc_ns.catalog.get_spec_source_file(ndt)
+        self.__register_dependent_types(spec, inc_ns, catalog, registered_types, in_progress_registrations)
+        if isinstance(spec, DatasetSpec):
+            built_spec = self.dataset_spec_cls.build_spec(spec)
+        else:
+            built_spec = self.group_spec_cls.build_spec(spec)
+        registered_types.add(ndt)
+        catalog.register_spec(built_spec, spec_file)
 
+    def __register_dependent_types(self, spec, inc_ns, catalog, registered_types, in_progress_registrations):
+        """Ensure that all types used by this type are registered.
+
+        This handles:
+        - Parent types (data_type_inc)
+        - Nested type definitions (data_type_def in child specs)
+        - Link target types
+
+        Circular dependencies are handled by checking in_progress_registrations before
+        recursing. This allows patterns like:
+        - Type A contains Type B, and Type B extends Type A
+        - Type A contains a reference to Type A (self-reference)
+        """
+        def __register_dependent_types_helper(child_spec):
+            if isinstance(child_spec, (GroupSpec, DatasetSpec)):
+                if child_spec.data_type_inc is not None:
+                    self.__register_type(
+                        child_spec.data_type_inc, inc_ns, catalog, registered_types, in_progress_registrations
+                    )
+                if child_spec.data_type_def is not None:  # nested type definition
+                    self.__register_type(
+                        child_spec.data_type_def, inc_ns, catalog, registered_types, in_progress_registrations
+                    )
+            else:  # spec is a LinkSpec
+                self.__register_type(
+                    child_spec.target_type, inc_ns, catalog, registered_types, in_progress_registrations
+                )
+            if isinstance(child_spec, GroupSpec):
+                for nested_spec in (child_spec.groups + child_spec.datasets + child_spec.links):
+                    __register_dependent_types_helper(nested_spec)
+
+        # Register parent type first
         if spec.data_type_inc is not None:
-            self.__register_type(spec.data_type_inc, inc_ns, catalog, registered_types)
+            self.__register_type(spec.data_type_inc, inc_ns, catalog, registered_types, in_progress_registrations)
+
+        # Register types from child specs (groups, datasets, links)
         if isinstance(spec, GroupSpec):
             for child_spec in (spec.groups + spec.datasets + spec.links):
-                __register_dependent_types_helper(child_spec, inc_ns, catalog, registered_types)
+                __register_dependent_types_helper(child_spec)
+
 
     @docval({'name': 'namespace_path', 'type': str, 'doc': 'the path to the file containing the namespaces(s) to load'},
             {'name': 'resolve',

--- a/src/hdmf/spec/namespace.py
+++ b/src/hdmf/spec/namespace.py
@@ -496,7 +496,7 @@ class NamespaceCatalog:
         :param in_progress_registrations: Set of types currently being registered (for circular dependency detection).
             None if the registration process is starting.
         """
-        if in_progress_registrations is None:  
+        if in_progress_registrations is None:
             in_progress_registrations = set()
         if ndt in registered_types or ndt in in_progress_registrations:
             # Already registered or currently being registered (circular dependency)

--- a/tests/unit/spec_tests/test_load_namespace.py
+++ b/tests/unit/spec_tests/test_load_namespace.py
@@ -548,3 +548,190 @@ class TestCoreExtensionConflicts(TestCase):
         for w in ws:
             self.assertNotIn("Schema conflict(s) detected in namespace 'extension'",
                              str(w.message))
+
+
+class TestCircularDependencies(TestCase):
+    """Test handling of circular type dependencies when loading namespaces.
+
+    This tests the fix for https://github.com/hdmf-dev/hdmf/issues/1364
+    """
+
+    def setUp(self):
+        self.tempdir = gettempdir()
+        self.core_source = 'core.yaml'
+        self.core_ns_path = 'core_namespace.yaml'
+        self.ext_source = 'extension.yaml'
+        self.ext_ns_path = 'extension_namespace.yaml'
+
+    def tearDown(self):
+        for f in (self.core_source, self.core_ns_path, self.ext_source, self.ext_ns_path):
+            remove_test_file(os.path.join(self.tempdir, f))
+
+    def test_circular_dependency_parent_contains_child_subtype(self):
+        """Test loading types where A contains B, and B extends A.
+
+        This is the pattern described in issue #1364:
+        - DynamicTable (A) contains a collection of MeaningsTable (B) objects
+        - MeaningsTable (B) is a subtype of DynamicTable (A)
+        """
+        # Create core namespace with ParentTable that contains ChildTable,
+        # and ChildTable extends ParentTable
+        parent_table_spec = GroupSpec(
+            'A parent table type',
+            data_type_def='ParentTable',
+            groups=[
+                GroupSpec(
+                    'A collection of child tables',
+                    data_type_inc='ChildTable',
+                    quantity='*',
+                )
+            ]
+        )
+        child_table_spec = GroupSpec(
+            'A child table that extends ParentTable',
+            data_type_def='ChildTable',
+            data_type_inc='ParentTable',
+        )
+
+        # Build and save core namespace
+        core_ns_builder = NamespaceBuilder('Core namespace', 'core', version='1.0.0')
+        core_ns_builder.add_spec(self.core_source, parent_table_spec)
+        core_ns_builder.add_spec(self.core_source, child_table_spec)
+        core_ns_builder.export(self.core_ns_path, outdir=self.tempdir)
+
+        # Load core namespace - this should not cause infinite recursion
+        ns_catalog = NamespaceCatalog()
+        ns_catalog.load_namespaces(os.path.join(self.tempdir, self.core_ns_path))
+
+        # Verify both types are registered
+        core_ns = ns_catalog.get_namespace('core')
+        registered_types = core_ns.get_registered_types()
+        self.assertIn('ParentTable', registered_types)
+        self.assertIn('ChildTable', registered_types)
+
+        # Verify the hierarchy is correct
+        hierarchy = ns_catalog.get_hierarchy('core', 'ChildTable')
+        self.assertEqual(hierarchy, ('ChildTable', 'ParentTable'))
+
+    def test_circular_dependency_from_included_namespace(self):
+        """Test loading circular dependencies from an included namespace.
+
+        This tests the specific code path in __load_namespace where types
+        are loaded from another namespace via the 'namespace' key in schema.
+        """
+        # Create core namespace with ParentTable and ChildTable
+        parent_table_spec = GroupSpec(
+            'A parent table type',
+            data_type_def='ParentTable',
+            groups=[
+                GroupSpec(
+                    'A collection of child tables',
+                    data_type_inc='ChildTable',
+                    quantity='*',
+                )
+            ]
+        )
+        child_table_spec = GroupSpec(
+            'A child table that extends ParentTable',
+            data_type_def='ChildTable',
+            data_type_inc='ParentTable',
+        )
+
+        core_ns_builder = NamespaceBuilder('Core namespace', 'core', version='1.0.0')
+        core_ns_builder.add_spec(self.core_source, parent_table_spec)
+        core_ns_builder.add_spec(self.core_source, child_table_spec)
+        core_ns_builder.export(self.core_ns_path, outdir=self.tempdir)
+
+        # Create extension namespace that includes types from core
+        ext_spec = GroupSpec(
+            'An extension type that uses ParentTable',
+            data_type_def='ExtType',
+            groups=[
+                GroupSpec(
+                    'A parent table',
+                    data_type_inc='ParentTable',
+                )
+            ]
+        )
+
+        ext_ns_builder = NamespaceBuilder('Extension namespace', 'extension', version='1.0.0')
+        ext_ns_builder.include_namespace('core')
+        ext_ns_builder.add_spec(self.ext_source, ext_spec)
+        ext_ns_builder.export(self.ext_ns_path, outdir=self.tempdir)
+
+        # Load core first, then extension
+        ns_catalog = NamespaceCatalog()
+        ns_catalog.load_namespaces(os.path.join(self.tempdir, self.core_ns_path))
+        ns_catalog.load_namespaces(os.path.join(self.tempdir, self.ext_ns_path))
+
+        # Verify types are registered in extension namespace
+        ext_ns = ns_catalog.get_namespace('extension')
+        registered_types = ext_ns.get_registered_types()
+        self.assertIn('ExtType', registered_types)
+        self.assertIn('ParentTable', registered_types)
+        self.assertIn('ChildTable', registered_types)
+
+    def test_self_referential_type(self):
+        """Test loading a type that contains itself (issue #794 pattern).
+
+        Example: A Container that can contain other Containers.
+        """
+        container_spec = GroupSpec(
+            'A container that can contain other containers',
+            data_type_def='RecursiveContainer',
+            groups=[
+                GroupSpec(
+                    'Nested containers',
+                    data_type_inc='RecursiveContainer',
+                    quantity='*',
+                )
+            ]
+        )
+
+        core_ns_builder = NamespaceBuilder('Core namespace', 'core', version='1.0.0')
+        core_ns_builder.add_spec(self.core_source, container_spec)
+        core_ns_builder.export(self.core_ns_path, outdir=self.tempdir)
+
+        # Load core namespace - this should not cause infinite recursion
+        ns_catalog = NamespaceCatalog()
+        ns_catalog.load_namespaces(os.path.join(self.tempdir, self.core_ns_path))
+
+        # Verify the type is registered
+        core_ns = ns_catalog.get_namespace('core')
+        registered_types = core_ns.get_registered_types()
+        self.assertIn('RecursiveContainer', registered_types)
+
+    def test_link_circular_dependency(self):
+        """Test circular dependency through links."""
+        # Create types where A links to B, and B extends A
+        parent_spec = GroupSpec(
+            'A parent type',
+            data_type_def='ParentType',
+            links=[
+                LinkSpec(
+                    name='child_link',
+                    doc='Link to a child type',
+                    target_type='ChildType',
+                )
+            ]
+        )
+        child_spec = GroupSpec(
+            'A child type that extends ParentType',
+            data_type_def='ChildType',
+            data_type_inc='ParentType',
+        )
+
+        core_ns_builder = NamespaceBuilder('Core namespace', 'core', version='1.0.0')
+        core_ns_builder.add_spec(self.core_source, parent_spec)
+        core_ns_builder.add_spec(self.core_source, child_spec)
+        core_ns_builder.export(self.core_ns_path, outdir=self.tempdir)
+
+        # Load core namespace - this should not cause infinite recursion
+        ns_catalog = NamespaceCatalog()
+        ns_catalog.load_namespaces(os.path.join(self.tempdir, self.core_ns_path))
+
+        # Verify both types are registered
+        core_ns = ns_catalog.get_namespace('core')
+        registered_types = core_ns.get_registered_types()
+        self.assertIn('ParentType', registered_types)
+        self.assertIn('ChildType', registered_types)

